### ISSUE TITLE
Can tispark query keep snapshot consistency?

### ISSUE THREAD

**Issue by chenatu** [pingcap/tispark#207]
Can tispark query keep snapshot consistency? What I mean is that, during the time when tispark executor scans on range of tikv, if some inserts or updates execute on the scanned and to-be-scanned rows, what category of consistency of data tispark will respond.

Another question is that, can we take a full snapshot of tidb data before we start tispark program and then tispark query on this prepared snapshot?

**Comment by zhexuany** [comment]
Hello there, 

Thanks for your interesting in using `TiSpark`. Currently, `Read Committed` and `Snapshot Isolation` are supported. As you may `Spark` is both cpu-bound and io-boud, we recommend you deploy spark and tikv separately. This may cause network latency but it can project you online business. 

For your second question, of course `tispark` can do it, but such snapshot need to reside in `tikv`. 

I hope above answer can help you. :)

**Comment by ilovesoup** [comment]
TiDB / TiKV store data in MVCC format and each transaction writes its change with version number (a timestamp basically). TiSpark's snapshot uses a timestamp as version number baseline. It will read all the data committed before the specified timestamp. The timestamp is generated by PD, as in TiDB. That is to say, both use the same global time. Even if when a transaction happened before snapshot time, as long as it does not commit, all the records it touches have lock flag on it, and TiSpark will skip those and searching for earlier version.
So basically, you read all the data already committed (Read committed isolation level)

**Comment by ilovesoup** [comment]
For 2nd question, yes. Each time you scan a table, it carries the timestamp you submit spark query. And this timestamp will give you a complete snapshot of data that committed already. You might cache it with Spark and do analyze on it.
The only exception to mention is: if a during a commit that TiDB crashed somehow, it might leave a transaction datasets locked but conceptually they are committed. TiDB will get those data by resolving lock but TiSpark for now will skip those. But still, remain read only committed level.
